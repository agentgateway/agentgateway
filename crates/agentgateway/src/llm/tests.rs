use std::fs;
use std::path::Path;

use serde::de::DeserializeOwned;

use super::*;

fn test_response<T: DeserializeOwned>(
	test_name: &str,
	xlate: impl Fn(T) -> Result<universal::ChatCompletionResponse, AIError>,
) {
	let test_dir = Path::new("src/llm/tests");

	// Read input JSON
	let input_path = test_dir.join(format!("{test_name}.json"));
	let provider_str = &fs::read_to_string(&input_path).expect("Failed to read input file");
	let provider_raw: Value =
		serde_json::from_str(provider_str).expect("Failed to parse provider json");
	let provider: T = serde_json::from_str(provider_str).expect("Failed to parse provider JSON");

	let openai_response =
		xlate(provider).expect("Failed to translate provider response to OpenAI format");

	insta::with_settings!({
			info => &provider_raw,
			description => input_path.to_string_lossy().to_string(),
			omit_expression => true,
			prepend_module_to_snapshot => false,
			snapshot_path => "tests",
	}, {
			 insta::assert_json_snapshot!(test_name, openai_response, {
			".id" => "[id]",
			".created" => "[date]",
		});
	});
}

fn test_request<T: Serialize>(
	provider_name: &str,
	test_name: &str,
	xlate: impl Fn(universal::ChatCompletionRequest) -> Result<T, AIError>,
) {
	let test_dir = Path::new("src/llm/tests");

	// Read input JSON
	let input_path = test_dir.join(format!("{test_name}.json"));
	let openai_str = &fs::read_to_string(&input_path).expect("Failed to read input file");
	let openai_raw: Value = serde_json::from_str(openai_str).expect("Failed to parse openai json");
	let openai: universal::ChatCompletionRequest =
		serde_json::from_str(openai_str).expect("Failed to parse openai JSON");

	let provider_response =
		xlate(openai).expect("Failed to translate OpenAI format to provider request ");

	insta::with_settings!({
			info => &openai_raw,
			description => format!("{}: {}", provider_name, test_name),
			omit_expression => true,
			prepend_module_to_snapshot => false,
			snapshot_path => "tests",
	}, {
			 insta::assert_json_snapshot!(format!("{}-{}", provider_name, test_name), provider_response, {
			".id" => "[id]",
			".created" => "[date]",
		});
	});
}

#[test]
fn test_bedrock() {
	let response = |i| bedrock::translate_response(i, &strng::new("fake-model"));
	test_response::<bedrock::types::ConverseResponse>("basic_bedrock", response);

	let request = |i| Ok(bedrock::translate_request(i));
	test_request("bedrock", "basic_input", request);
	test_request("bedrock", "full_input", request);
}

#[test]
fn test_anthropic() {
	let response = |i| Ok(anthropic::translate_response(i));
	test_response::<anthropic::types::MessagesResponse>("basic_anthropic", response);

	let request = |i| Ok(anthropic::translate_request(i));
	test_request("anthropic", "basic_input", request);
	test_request("anthropic", "full_input", request);
}
