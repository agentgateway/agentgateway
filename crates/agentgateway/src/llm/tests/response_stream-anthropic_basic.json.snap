---
source: crates/agentgateway/src/llm/tests.rs
description: src/llm/tests/response_stream-anthropic_basic.json
---
event: 
data: {"id":"msg_016wbjv5k86nxxd8CEZwSQnA","choices":[{"index":0,"delta":{"content":"Hi there! How are"}}],"created":123,"model":"claude-3-5-haiku-20241022","service_tier":null,"system_fingerprint":null,"object":"chat.completion.chunk","usage":null}

event: 
data: {"id":"msg_016wbjv5k86nxxd8CEZwSQnA","choices":[{"index":0,"delta":{"content":" you doing today?"}}],"created":123,"model":"claude-3-5-haiku-20241022","service_tier":null,"system_fingerprint":null,"object":"chat.completion.chunk","usage":null}

event: 
data: {"id":"msg_016wbjv5k86nxxd8CEZwSQnA","choices":[{"index":0,"delta":{"content":" Is"}}],"created":123,"model":"claude-3-5-haiku-20241022","service_tier":null,"system_fingerprint":null,"object":"chat.completion.chunk","usage":null}

event: 
data: {"id":"msg_016wbjv5k86nxxd8CEZwSQnA","choices":[{"index":0,"delta":{"content":" there anything I can help"}}],"created":123,"model":"claude-3-5-haiku-20241022","service_tier":null,"system_fingerprint":null,"object":"chat.completion.chunk","usage":null}

event: 
data: {"id":"msg_016wbjv5k86nxxd8CEZwSQnA","choices":[{"index":0,"delta":{"content":" you with?"}}],"created":123,"model":"claude-3-5-haiku-20241022","service_tier":null,"system_fingerprint":null,"object":"chat.completion.chunk","usage":null}

event: 
data: {"id":"msg_016wbjv5k86nxxd8CEZwSQnA","choices":[],"created":123,"model":"claude-3-5-haiku-20241022","service_tier":null,"system_fingerprint":null,"object":"chat.completion.chunk","usage":{"prompt_tokens":15,"completion_tokens":21,"total_tokens":36}}
